Prefer extension renderers to core renderers

rickysixx opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://docs.jellyfin.org/general/contributing/issues.html page.
-->

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->
Prefer extension renderers for ExoPlayer to core renderers.

Extension renderes like `LibflacAudioRenderer` and `LibopusAudioRenderer` seem to work better than the core ones.

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->
Fixes #3816

nielsvanvelzen commented:
We should generally prefer the media3 implementation of codecs as those can use platform decoders, including ones made by vendors. This will, at times, be much more efficient then using the ffmpeg renderer. Media3 should only use the ffmpeg renderer when there is no platform decoder or the platform decoder is not compatible.

rickysixx commented:
@nielsvanvelzen Unfortunately it seems that media3 fallback to ffmpeg renderers does not (always) work properly, at least in my case (see #3816).

I have no experience with media playback on Android so I cannot tell if this PR is the proper fix to #3816, but I have tested it on different movies/shows and it seem to work fine. I still feel that this is somewhat a quick & dirty hack, but I was not able to come up with a different non-hacky solution.

Maybe we could add an advanced preference to allow the user to switch between `EXTENSION_RENDERER_MODE_ON` and `EXTENSION_RENDERER_MODE_PREFER` (the default value would be `EXTENSION_RENDERER_MODE_ON`, as it is right now). In this way the user can at least try to play the media anyway instead of being stuck due to playback errors with media3 renderers.

nielsvanvelzen commented:
The linked issue does not contain full logs so I can't really tell much about it. Normally the player should fallback to the next decoder when something is wrong.

zkhcohen commented:
> The linked issue does not contain full logs so I can't really tell much about it. Normally the player should fallback to the next decoder when something is wrong.

Could you provide a little more insight about how this decoder-fallback mechanism is implemented? I think it might help with troubleshooting the behavior I'm seeing in the following comment, specifically how media may begin to play after a few minutes:

https://github.com/jellyfin/jellyfin-androidtv/issues/3830#issuecomment-2276891040

nielsvanvelzen reviewed (changes requested):
I'd be fine introducing this behavior under a developer preference. It would be called something like "Prefer FFmpeg for audio playback".

Additionally it should also propagate to the new playback engine.

rickysixx commented:
I have added the new preference.

I have also added a check to activate [ExoPlayer debug logs](https://developer.android.com/media/media3/exoplayer/debug-logging) which help to understand which decoder is being used.

I am however facing some troubles to propagate this change to the new playback engine (e.g. `org.jellyfin.playback.media3.exoplayer.ExoPlayerBackend`, if I am not wrong). Using the `UserPreferences` class there would introduce a circular dependency between modules `jellyfin-androidtv.app.main` and `jellyfin-androidtv.playback.media3.exoplayer.main`.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/preference/UserPreferences.kt:
```diff
@@ -195,6 +195,11 @@ class UserPreferences(context: Context) : SharedPreferenceStore(
+		/**
+		 * Whether to prefer extension renderers (i.e. FFmpeg) to core ones.
+		 */
+		var preferExtensionRenderers = booleanPreference("prefer_extension_renderers", defaultValue = false)
```
```suggestion
		var preferExoPlayerFfmpeg = booleanPreference("exoplayere_prefer_ffmpeg", false)
```

nielsvanvelzen replied:
Also move it up to after refreshRateSwitchingBehavior so it's grouped with the other playback preferences.

rickysixx replied:
done

nielsvanvelzen commented on app/src/main/res/values/strings.xml:
```diff
@@ -514,6 +514,8 @@
     <string name="scheduled_in_next_24_hours">Scheduled in next 24 hours</string>
     <string name="past_24_hours">Past 24 hours</string>
+    <string name="pref_prefer_extension_renderers_label">Prefer FFmpeg for audio playback</string>
+    <string name="pref_prefer_extension_renderers_description">May fix some playback errors and avoid audio transcoding.</string>
```
```suggestion
    <string name="pref_prefer_extension_renderers_description">Use FFMpeg to decode audio codecs when platform codecs are available.</string>
```

nielsvanvelzen reviewed (changes requested):
To propagate the preference to the ExoPlayerBackend you'll have to add a new boolean to ExoPlayerOptions class and populate it from the PlaybackModule where we initialize the playback code.

This does mean it requires an app restart when the preference changes, but that's fine for now.

rickysixx commented:
Requested changes applied.

Thank you for your help with the code!

nielsvanvelzen reviewed (approved):
Thanks! I've pushed a final commit with some small changes.
